### PR TITLE
refactor(web): subkey abstraction (step 1) - subkey selection 'returned' via Promise

### PR DESCRIPTION
**Pull request by jahorton**
A stepping-stone toward #5023 and the first part of an arc aimed at refactoring how the on-screen keyboard longpress gestures (both on websites and when embedded within apps) are handled.

There are three main sections to this description:
1. Arc overview
2. Future goals
3. For this PR only

------

### Arc overview

1. #5294
    - You are here.
    - Step 1 theme:  subkey commands as async events (as _gestures_ are inherently async)
2. #5295
3. #5354
    - Step 2 theme:  code path cleanup / clarification
4. #5355
5. #5356
    - Step 3 theme:  prep for common abstraction for the two control flows
6. #5357
    - that's right - "Gesture".  These aren't intended for _just_ longpresses, though they'll be the only supported gesture at first.
    - As gestures are events that evaluate over intervals of time, they are inherently asynchronous.  Hence, the use of `Promise`s by earlier PRs in the chain.
7. #5358
    - Given a "gesture" abstraction, establishes gesture-oriented patterns 

------

At this stage, the following content from #5357 (PR 6 of 7) may prove of interest and/or motivating:

https://github.com/keymanapp/keyman/blob/6034f252e70c2218a1983059324f32d2bdf15daa/web/source/osk/pendingGesture.interface.ts#L2-L8

`PendingGesture` - can model a longpress gesture that has not yet been confirmed, thus does not block base-key interactions / change of highlighted key

https://github.com/keymanapp/keyman/blob/6034f252e70c2218a1983059324f32d2bdf15daa/web/source/osk/realizedGesture.interface.ts#L2-L9

`RealizedGesture` - can model a _confirmed_ longpress gesture that has displayed subkeys but not yet completed (as in, hasn't yet returned a subkey)

------

### Future goals

The abstractions from the previous block also provide a starting point to, in the future, model LDML's [flick](https://unicode-org.github.io/cldr/ldml/tr35-keyboards.html#Element_flicks) and the [multitap attribute](https://unicode-org.github.io/cldr/ldml/tr35-keyboards.html#Element_map).
- There's nothing saying a gesture can't also be automatically 'realized' / resolved.
    - I would imagine this to be useful for `flick` events.  `Promise`s can be built pre-resolved, after all.
- Once one potential gesture is 'realized' for a touch event, the others ought be cancelled.
    - `multitap` may make this a bit more complex, but it'll be manageable via some sort of state machine if so.
- `multitap` doesn't clear for new touches (on the same base key) when others would, as it is a gesture across multiple touches.
    - Still, it can simply be left 'pending' until the confirming second tap is released, which would then auto-cancel flicks, longpresses, and (new, non-multitap) output from the base key while taps continue to maintain the new 'realized' / 'active' state.
    - Simillarly, a confirmed flick could auto-cancel potential multitaps, which (approximately) maintain the same touch coordinate.
    - Trickiest part:  if the output should "rotate" as multitaps continue, the `RealizedGesture.promise` typing may need a minor rework, as `Promise`s may only resolve once.  That's a bridge we can cross when we come to it, though.
- So, having a common abstraction for all gestures simplifies tracking all gestures for a touch event & for mass-cancellation of the gestures that didn't "win".

We'll probably want some sort of formalized "gesture engine" before implementing either.  Design pending, of course, though at least the new abstractions will greatly facilitate that process.

------

### For this PR only

The current pathways for processing subkeys - one for 'in-browser' contexts, one for 'embedded' contexts - are quite different in their requirements and operational patterns.  Up to this point, KMW has implemented the 'embedded' version by a classic JS 'extension' technique, overriding the prototype's version of relevant methods with new ones.  This pattern will be... less easily maintainable as we further modularize the OSK into its own, independent module.

As gestures (such as longpresses for subkeys) are events that occur over a time interval, they are inherently asynchronous... hence this PR's shift toward use of `Promise`s.

This PR focuses on _part_ of the process of properly abstracting subkey display and handling.  The core tenets for such an abstraction:
- **Input**: the `VisualKeyboard` notes that a subkey array should be displayed to the user based upon a selected subkey.
- **Output**: for dismissal, the subkey array will either:
    - be cancelled (i.e, dismissed - returning no event)
    - return a key event corresponding to one of the displayed subkeys

From the perspective of whatever code displays the subkey array, the desired output occurs asynchronously.  Up to now, the two pathways have directly applied the subkey's event to the current context.

The big, motivating factor:  for the OSK-Core, the goal is for the OSK to 'raise' the event for consumption by the greater engine at large, rather than directly applying it.  For this to work well for the 'embedded' OSK pathway, embedded subkeys must then be passed through the OSK's event-raising mechanism (which is currently not the case).  Ideally, the OSK-Core will be able to implement much of the code required to integrate our current 'embedded' behavior.

Requiring use of `Promise`s as part of the eventual subkey-behavior abstraction will facilitate event-raising for subkeys (making it _far_ more easily centralized) explicitly.  This also works well with the explicitly-asynchronous subkey selection that is currently required when embedded within Android.

### CODE REVIEW

File: web/source/osk/visualKeyboard.ts [hunk 1]
```diff
@@ -60,13 +65,12 @@ namespace com.keyman.osk {
...
     // Popup key management
-    popupBaseKey: KeyElement;
-    popupPending: boolean = false;
-    subkeyDelayTimer: number;
     popupDelay: number = 500;
+    subkeyDelayTimer: number;
     menuEvent: KeyElement; // Used by embedded-mode.
     keytip: KeyTip;
     subkeyPopup: browser.SubkeyPopup;
+    subkeyDelegator: any; // a temp, intermediate property during subkey abstraction work
```
Comment (jahorton) [inline web/source/osk/visualKeyboard.ts:73]: Extra comment for reviewers:  this is an 'embedded' mode parallel to the `browser.SubkeyPopup` class.

As `SubkeyPopup` manages in-browser ('native-mode') subkey selection, the "subkey delegator" introduced here is used to "manage" (integrate) embedded-mode subkey selection.

Comment (jahorton) [reply web/source/osk/visualKeyboard.ts:73]: This introduces the parallel now; later PRs in the chain will bring them closer together and lead to a point where an abstract interface for both may be defined.

File: web/source/kmwembedded.ts [hunk 2]
```diff
@@ -60,8 +61,19 @@ namespace com.keyman.osk {
...
+        let promise = new Promise<text.KeyEvent>(function(resolve, reject) {
+          let delegator = new embedded.SubkeyDelegator(key, resolve);
+          _this.subkeyDelegator = delegator;
+        }).then(function(keyEvent) {
+          _this.subkeyDelegator = null;
+          // Allow active cancellation, even if the source should allow passive.
+          // It's an easy and cheap null guard.
+          if(keyEvent) {
+            PreProcessor.raiseKeyEvent(keyEvent);
+          }
```
Comment (mcdurdin) [inline web/source/kmwembedded.ts:74]: I don't understand this comment, nor how it relates to the associated code

File: web/source/kmwembedded.ts [hunk 3]
```diff
@@ -60,8 +61,19 @@ namespace com.keyman.osk {
 
         // #3718: No longer prepend base key to subkey array
 
-        this.popupBaseKey = key;
-        this.popupPending=true;
+        let _this = this;
+        let promise = new Promise<text.KeyEvent>(function(resolve, reject) {
```
Comment (mcdurdin) [inline web/source/kmwembedded.ts:65]: Does `reject` parameter need to be included here if it is not used?

File: web/source/kmwembedded.ts [hunk 4]
```diff
@@ -349,11 +361,13 @@ namespace com.keyman.text {
       keymanweb.domManager.initActiveElement(Lelem);
 
       var nextLayer: string;
-
+      var delegator: com.keyman.osk.embedded.SubkeyDelegator = null;
```
Comment (mcdurdin) [inline web/source/kmwembedded.ts:364]: It'd be really helpful if the purpose of the `SubkeyDelegator` was documented

File: web/source/osk/embedded/subkeyDelegator.ts [hunk 5]
```diff
@@ -0,0 +1,19 @@
+namespace com.keyman.osk.embedded {
+  export class SubkeyDelegator {
```
Comment (mcdurdin) [inline web/source/osk/embedded/subkeyDelegator.ts:2]: What's the difference between the `SubkeyDelegator` and the `SubkeyPopup`? Why are both needed? `SubkeyDelegator` seems to be a subset of `SubkeyPopup`?

File: web/source/osk/visualKeyboard.ts [hunk 6]
```diff
@@ -630,10 +640,15 @@ namespace com.keyman.osk {
       // Process and clear highlighting of pending target
       if(this.keyPending) {
         this.highlightKey(this.keyPending,false);
-
+
```
Comment (mcdurdin) [inline web/source/osk/visualKeyboard.ts:643]: It'd be good if you could auto-trim whitespace in your editor.

File: web/source/osk/visualKeyboard.ts [hunk 7]
```diff
@@ -727,11 +743,11 @@ namespace com.keyman.osk {
         return;
       }
 
-      var sk=document.getElementById('kmw-popup-keys');
+      var sk=this.subkeyPopup
```
Comment (mcdurdin) [inline web/source/osk/visualKeyboard.ts:746]:
```suggestion
      var sk=this.subkeyPopup;
```

File: web/source/osk/visualKeyboard.ts [hunk 8]
```diff
@@ -1680,19 +1694,38 @@ namespace com.keyman.osk {
    *
    * @param   {Object}  key   base key object
    */
-  touchHold(key: KeyElement) {
+  touchHold(key: KeyElement, force?: boolean) {
```
Comment (mcdurdin) [inline web/source/osk/visualKeyboard.ts:1697]: What is `force` for? Can you document it above? When is force=false?

File: web/source/osk/visualKeyboard.ts [hunk 9]
```diff
@@ -1680,19 +1694,38 @@ namespace com.keyman.osk {
...
+            _this.showSubKeys(key, resolve);
+          }, force ? 0 : _this.popupDelay);
+
+          _this.subkeyDelayTimer = timerId;
+      }).then(function(keyEvent: text.KeyEvent) {
+        // Allow active cancellation, even if the source should allow passive.
+        // It's an easy and cheap null guard.
+        if(keyEvent) {
+          PreProcessor.raiseKeyEvent(keyEvent);
+        }
```
Comment (mcdurdin) [inline web/source/osk/visualKeyboard.ts:1727]: Same as before; I don't understand this comment. `It's an easy and cheap null guard` is not really very helpful in terms of understanding the why.

File: web/source/osk/visualKeyboard.ts [hunk 10]
```diff
@@ -1680,19 +1694,38 @@ namespace com.keyman.osk {
...
-      this.subkeyDelayTimer = window.setTimeout(
-        function(this: VisualKeyboard) {
-          this.clearPopup();
-          this.showSubKeys(key);
-        }.bind(this), this.popupDelay);
+      let _this = this;
+
+      // This Promise receives the user's selected subkey, should it exist.
+      // Note:  the OSK's use of this Promise will allow for passive cancellation.
+      //        If cancelled (so, `null`), the Promise ought remain unresolved.
```
Comment (mcdurdin) [inline web/source/osk/visualKeyboard.ts:1710]: What does `null` refer to here?

**Review by mcdurdin — COMMENTED** [review]
There are a lot of structural changes here that I don't really understand. It's pretty hard to parse out the reason for many of the changes, despite the commentary in the PR. I would like to see a lot more code commenting as the abstraction makes it pretty hard to follow the flow of execution, which seems to happen across 3 or 4 different files.

This is going to need a lot of regression testing -- there are significant changes in behaviour and it's going to impact both embedded and native mode.

File: web/source/osk/embedded/subkeyDelegator.ts (hunk 5 above)
Comment (jahorton) [reply web/source/osk/embedded/subkeyDelegator.ts:2]: With regard to the description, `SubkeyDelegator` becomes the embedded implementation of `RealizedGesture` for subkeys, while `SubkeyPopup` becomes the in-browser (site-integrated) implementation of `RealizedGesture`.

Comment (mcdurdin) [reply web/source/osk/embedded/subkeyDelegator.ts:2]: Can you add comments to the code accordingly? I understand the names now but I had no idea what `SubkeyDelegator` was delegating to at first 😆

**Comment by jahorton** [comment]
Abandoning in favor of #5387 - a documented + combined version of all seven, as many of the changes in later steps reshuffle code from earlier steps.
